// @framer/nvh.isometric/code/Isometric.tsx
import * as React from "react";
import { Frame, FramerEvent, PropertyControls, ControlType } from "framer";

interface Props {
  rotationX: number;
  rotationY: number;
  rotationZ: number;
  backgroundColor: string;
}

export class Isometric extends React.Component<Props> {
  static defaultProps = {
    rotationX: 50,
    rotationY: 0,
    rotationZ: 320
  };

  static propertyControls: PropertyControls<Props> = {
    rotationX: {
      type: ControlType.Number,
      title: "Rotation X",
      min: 0,
      max: 360,
      round: false
    },
    rotationY: {
      type: ControlType.Number,
      title: "Rotation Y",
      min: 0,
      max: 360
    },
    rotationZ: {
      type: ControlType.Number,
      title: "Rotation Z",
      min: 0,
      max: 360
    }
  };

  render() {
    const { width, height } = this.props;
    return (
      <Frame
        width={width}
        height={height}
        rotationX={this.props.rotationX}
        rotationY={this.props.rotationY}
        rotationZ={this.props.rotationZ}
        scale={0.6}
        background={null}
      >
        {this.props.children}
      </Frame>
    );
  }
}
